src/L1/rollup/ScrollChain.sol

```diff
@@ -438,149 +369,100 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
-        bytes32 _publicInputHash;
-        {
-            bytes32 _dataHash = BatchHeaderV0Codec.getDataHash(batchPtr);
-            bytes32 _blobVersionedHash = BatchHeaderV1Codec.getBlobVersionedHash(batchPtr);
-            bytes32 _prevStateRoot = finalizedStateRoots[_batchIndex - 1];
-            // verify blob versioned hash
-            _checkBlobVersionedHash(_blobVersionedHash, _blobDataProof);
-            _publicInputHash = keccak256(
-                abi.encodePacked(
-                    layer2ChainId,
-                    _prevStateRoot,
-                    _postStateRoot,
-                    _withdrawRoot,
-                    _dataHash,
-                    _blobDataProof[0:64],
-                    _blobVersionedHash
-                )
-            );
-        }
-
-        // load version from batch header, it is always the first byte.
-        uint256 batchVersion = BatchHeaderV0Codec.getVersion(batchPtr);
-        // verify batch
-        IRollupVerifier(verifier).verifyAggregateProof(batchVersion, _batchIndex, _aggrProof, _publicInputHash);
-
-        // Pop finalized and non-skipped message from L1MessageQueue.
-        uint256 _totalL1MessagesPoppedOverall = BatchHeaderV0Codec.getTotalL1MessagePopped(batchPtr);
-        _popL1MessagesMemory(
-            BatchHeaderV1Codec.getSkippedBitmapPtr(batchPtr),
-            _totalL1MessagesPoppedOverall,
-            BatchHeaderV0Codec.getL1MessagePopped(batchPtr)
-        );
-
-        _afterFinalizeBatch(_totalL1MessagesPoppedOverall, _batchIndex, _batchHash, _postStateRoot, _withdrawRoot);
+    /// @dev This function will only allow security council to call once.
+    function finalizeEuclidInitialBatch(bytes32 postStateRoot) external override onlyOwner {
+        if (postStateRoot == bytes32(0)) revert ErrorStateRootIsZero();
+
+        uint256 batchIndex = initialEuclidBatchIndex;
+        // make sure only finalize once
+        if (finalizedStateRoots[batchIndex] != bytes32(0)) revert ErrorBatchIsAlreadyVerified();
+        // all v4 batches should be finalized
+        if (lastFinalizedBatchIndex + 1 != batchIndex) revert ErrorNotAllV4BatchFinalized();
+
+        // update storage
+        lastFinalizedBatchIndex = batchIndex;
+        bytes32 withdrawRoot = withdrawRoots[batchIndex - 1];
+        finalizedStateRoots[batchIndex] = postStateRoot;
+        withdrawRoots[batchIndex] = withdrawRoot;
+
+        emit FinalizeBatch(batchIndex, committedBatches[batchIndex], postStateRoot, withdrawRoot);
     }
 
     /// @inheritdoc IScrollChain
-    function finalizeBundleWithProof(
-        bytes calldata _batchHeader,
-        bytes32 _postStateRoot,
-        bytes32 _withdrawRoot,
-        bytes calldata _aggrProof
+    /// @dev This function only finalize batches version >= 6. If prover try to finalize v4, v5, v6 batches
+    /// together, the verifier should revert. So we don't do any extra checks in this function.
+    function finalizeBundlePostEuclid(
```

Comment:
Can't we keep the same public input format and continue using `finalizeBundleWithProof` (remove `finalizeBundlePostEuclid`)? Seems unnecessary to switch to a new functions that's basically the same.

I think you can just keep the same public input and ignore `numBatches` inside `ZkEvmVerifierPostEuclid`.

Reply:
Updated in [0fd1107](https://github.com/scroll-tech/scroll-contracts/pull/73/commits/0fd1107dbf39a2a06d9736f8b97f911485dccffa).